### PR TITLE
Add global state

### PR DESCRIPTION
**Pull request by kaspatel-mdsol**
This PR does the following things.

- Add Typescript
- Add Prettier pre-commit hook that automatically formats files
- Add Global State Context so you can have sharable state across components

### CODE REVIEW

File: components/StateContext.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
+import React, { useReducer, useContext, Dispatch, ReactNode } from 'react';
+import { DesignerSchemaType, DesignerActionType } from '../types';
+
+interface IDesignerSchema {
```
Comment (ncarroll-mdsol) [inline components/StateContext.tsx:4]: How do all these interfaces get used? Is there a good guide for this sort of global state?

File: components/StateContext.tsx [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+interface IDesignerSchemaMap {
+  [key: string]: IDesignerSchema;
+}
+
+interface IDesignerScreenMap {
+  [key: string]: IDesignerScreen;
+}
+
+interface IDesignerState {
+  schemas: IDesignerSchemaMap;
```
Comment (ncarroll-mdsol) [inline components/StateContext.tsx:26]: What are the schemas in this case?

File: components/WorkflowBuilderPage.tsx [hunk 3]
```diff
@@ -33,9 +35,9 @@ const WorkflowBuilderPage = () => {
   node2.current.setPosition(400, 100);
 
   // link the ports
-  const link1 = useRef(port1.current.link(port2.current));
+  const link1 = useRef(port1.current.link<DefaultLinkModel>(port2.current));
   link1.current.getOptions().testName = 'Test';
-  link1.current.addLabel('Hello World!');
+  link1.current.addLabel('Hello World');
```
Comment (ncarroll-mdsol) [inline components/WorkflowBuilderPage.tsx:40]: ??? You don't like the exclamation point?? 😂

File: components/StateContext.tsx (hunk 1 above)
Comment (kaspatel-mdsol) [reply components/StateContext.tsx:4]: This is a WIP schema type. A loose defn of a schema in my mind is "type of element that covers 100% of width on a screen aka `A Line Item`". 

Based on several different schema configs, we should be able to render different screens.

Currently, I have provision for Text, Image, and Select types. Ideally, we would have more fields in each schema describing a few more things about it, along with interfaces for each of these schema type and a union of Schema Type that will encompass all different schema/line item possible on a screen.

File: components/StateContext.tsx (hunk 2 above)
Comment (kaspatel-mdsol) [reply components/StateContext.tsx:26]: If you check IDesignerSchemaMap, you'll see it basically is an object/hash/map of string(id) to a particular schema.

File: components/WorkflowBuilderPage.tsx (hunk 3 above)
Comment (kaspatel-mdsol) [reply components/WorkflowBuilderPage.tsx]: aah.. my mistake. while converting to typescript I messed up. Reverted

File: components/StateContext.tsx (hunk 1 above)
Comment (kaspatel-mdsol) [reply components/StateContext.tsx:4]: Also, if you think a line item is a better name, I can change that. I feel LineItem is better than schema, only thought of it while typing the above comment.

Comment (ncarroll-mdsol) [reply components/StateContext.tsx:4]: I think Schema is fine for now but we can always change it later. In some of the recent designer conversations with product we've been using the term Control for an element that gets put on the page. And a Container for a full-width or full-height view, which can contain multiple controls.

**Pull request merged**